src/util/ALE.sol

```diff
@@ -0,0 +1,235 @@
+        address _buyTokenAddress,
+        // The `allowanceTarget` field from the API response.
+        address _spender,
+        // The `data` field from the API response.
+        bytes calldata _swapCallData,
+        // Permit data
+        Permit calldata _permit,
+        // Optional helper data in case the collateral needs to be transformed
+        bytes calldata _helperData
+    ) external payable nonReentrant {
+        if (markets[_buyTokenAddress].market == address(0))
+            revert MarketNotSetForCollateral(_buyTokenAddress);
+
+        IMarket market = IMarket(markets[_buyTokenAddress].market);
+
+        // Mint and approve
+        dola.mint(address(this), _value);
+        dola.approve(_spender, _value);
+
+        IERC20 buyToken = IERC20(_buyTokenAddress);
+        uint256 collateralBalBefore = buyToken.balanceOf(address(this));
+        // Call the encoded swap function call on the contract at `swapTarget`,
+        // passing along any ETH attached to this function call to cover protocol fees.
+        (bool success, ) = exchangeProxy.call{value: msg.value}(_swapCallData);
+        if (!success) revert SwapFailed();
+
+        // Actual collateral/buyToken bought
+        uint256 collateralAmount = buyToken.balanceOf(address(this))
+        - collateralBalBefore;
+
+        // If there's an helper contract, the buyToken has to be transformed
+        if (markets[_buyTokenAddress].helper != address(0)) {
+            // Approve collateral to be transformed
+            buyToken.approve(
+                markets[_buyTokenAddress].helper,
+                collateralAmount
+            );
+            // Collateral amount is now transformed
+            collateralAmount = ITransformHelper(
+                markets[_buyTokenAddress].helper
+            ).transformToCollateral(collateralAmount, _helperData);
+            // Approve market to spend transformed collateral
+            IERC20(markets[_buyTokenAddress].collateral).approve(
+                address(market),
+                collateralAmount
+            );
+        } else {
+            // If there's no helper it means we already bought the collateral token so we can deposit
+            // Approve market to spend collateral
+            buyToken.approve(address(market), collateralAmount);
+        }
+
+        // Deposit and borrow on behalf
+        market.deposit(msg.sender, collateralAmount);
+
+        // We borrow the amount of DOLA we minted before
+        market.borrowOnBehalf(
+            msg.sender,
+            _value,
+            _permit.deadline,
+            _permit.v,
+            _permit.r,
+            _permit.s
+        );
+
+        if (dola.balanceOf(address(this)) < _value) revert DOLAInvalidBorrow();
+
+        // Burn the dola minted previously
+        dola.burn(_value);
+
+        // Refund any possible unspent 0x protocol fees to the sender.
+        payable(msg.sender).transfer(address(this).balance);
```

Comment:
Not sure how often there'd be leftover funds, but this may be an optimization
```suggestion
        // Refund any possible unspent 0x protocol fees to the sender.
        if(address(this).balance > 0) payable(msg.sender).transfer(address(this).balance);
```

Reply:
Fixed in [21265a2](https://github.com/InverseFinance/FiRM/pull/52/commits/21265a282ef6b19fa53c46e5b48fe83f48880686)